domain/src/main/java/org/oppia/domain/topic/TopicController.kt

```diff
@@ -191,6 +245,215 @@ class TopicController @Inject constructor(
+                storyBuilder.setChapter(chapterIndex, chapterBuilder)
+              }
+            }
+          }
+          topicBuilder.setStory(storyIndex, storyBuilder.build())
+        } else {
+          if (storySummary.chapterList.isNotEmpty()) {
+            val updatedStorySummary = setFirstChapterAsNotStarted(storySummary)
+            topicBuilder.setStory(storyIndex, updatedStorySummary)
+          }
+        }
+      }
+    } else {
+      topic.storyList.forEachIndexed { storyIndex, storySummary ->
+        if (storySummary.chapterList.isNotEmpty()) {
+          val updatedStorySummary = setFirstChapterAsNotStarted(storySummary)
+          topicBuilder.setStory(storyIndex, updatedStorySummary)
+        }
+      }
+    }
+    return topicBuilder.build()
+  }
+
+  /** Combines the specified story-summary without progress and story-progress into a new topic. */
+  private fun combineStorySummaryAndStoryProgress(
+    storySummary: StorySummary,
+    storyProgress: StoryProgress
+  ): StorySummary {
+    if (storyProgress.chapterProgressMap.isNotEmpty()) {
+      val storyBuilder = storySummary.toBuilder()
+      storySummary.chapterList.forEachIndexed { chapterIndex, chapterSummary ->
+        if (storyProgress.chapterProgressMap.containsKey(chapterSummary.explorationId)) {
+          val chapterBuilder = chapterSummary.toBuilder()
+          chapterBuilder.chapterPlayState = ChapterPlayState.COMPLETED
+          storyBuilder.setChapter(chapterIndex, chapterBuilder)
+        } else {
+          if (storyBuilder.getChapter(chapterIndex - 1).chapterPlayState == ChapterPlayState.COMPLETED) {
+            val chapterBuilder = chapterSummary.toBuilder()
+            chapterBuilder.chapterPlayState = ChapterPlayState.NOT_STARTED
+            storyBuilder.setChapter(chapterIndex, chapterBuilder)
+          } else {
+            val chapterBuilder = chapterSummary.toBuilder()
+            chapterBuilder.chapterPlayState = ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES
+            storyBuilder.setChapter(chapterIndex, chapterBuilder)
+          }
+        }
+      }
+      return storyBuilder.build()
+    } else {
+      return setFirstChapterAsNotStarted(storySummary)
+    }
+  }
+
+  // TODO(#21): Expose this as a data provider, or omit if it's not needed.
+  internal fun retrieveTopic(topicId: String): Topic {
+    return when (topicId) {
+      TEST_TOPIC_ID_0 -> createTestTopic0()
+      TEST_TOPIC_ID_1 -> createTestTopic1()
+      FRACTIONS_TOPIC_ID -> createTopicFromJson(
+        "fractions_topic.json", "fractions_skills.json", "fractions_stories.json"
+      )
+      RATIOS_TOPIC_ID -> createTopicFromJson(
+        "ratios_topic.json", "ratios_skills.json", "ratios_stories.json"
+      )
+      else -> throw IllegalArgumentException("Invalid topic ID: $topicId")
+    }
+  }
+
+  private fun retrieveTopicAsync(topicId: String): AsyncResult<Topic> {
```

Comment:
Why is this needed? Can't we just use retrieveTopic()? DataProviders will catch the exception and turn it into an AsyncResult.failed().

Reply:
Okay, Actually I was not sure that it will turn into AsyncResult.failed() or not, but now as I run test cases, it makes sure that it does work therefore removing both of these Async functions.